File: fetch-earnings.js
#!/usr/bin/env node

/**
 * Script to fetch earnings data for all companies
 * 
 * RECOMMENDED API: Financial Modeling Prep (FMP)
 * - Free tier: 250 API calls per day (enough for most updates)
 * - Get free key at: https://site.financialmodelingprep.com/developer/docs
 * 
 * Usage:
 *   export FMP_API_KEY="your_key_here"
 *   node fetch-earnings.js
 * 
 * For large datasets, you can run this script over multiple days.
 * The script tracks which companies already have earnings and skips them.
 */

const fs = require('fs');
const https = require('https');
const path = require('path');

// Load .env file if it exists
function loadEnv() {
    const envPath = path.join(__dirname, '.env');
    if (fs.existsSync(envPath)) {
        const envFile = fs.readFileSync(envPath, 'utf8');
        envFile.split('\n').forEach(line => {
            line = line.trim();
            if (line && !line.startsWith('#') && line.includes('=')) {
                const [key, ...valueParts] = line.split('=');
                const value = valueParts.join('=');
                process.env[key] = value;
            }
        });
    }
}

loadEnv();

// Load existing data
const capexData = JSON.parse(fs.readFileSync('./data/capex_data.json', 'utf8'));

// API Keys (loaded from .env file)
const FMP_API_KEY = process.env.FMP_API_KEY || '';
const ALPHA_VANTAGE_API_KEY = process.env.ALPHA_VANTAGE_API_KEY || '';

// Track API usage
let apiCallCount = 0;
const MAX_CALLS_PER_MINUTE = 10; // Rate limiting
const companiesProcessed = [];
const companiesFailed = [];

/**
 * Make HTTPS GET request
 */
function httpsGet(url) {
    return new Promise((resolve, reject) => {
        https.get(url, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                if (res.statusCode === 200) {
                    try {
                        resolve(JSON.parse(data));
                    } catch (e) {
                        reject(new Error('Failed to parse JSON response'));
                    }
                } else {
                    reject(new Error(`HTTP ${res.statusCode}: ${data}`));
                }
            });
        }).on('error', reject);
    });
}

/**
 * Sleep helper for rate limiting
 */
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Fetch earnings from Financial Modeling Prep (RECOMMENDED - Most reliable)
 * Free tier: 250 calls/day
 * https://site.financialmodelingprep.com/developer/docs
 */
async function fetchEarningsFromFMP(symbol) {
    if (!FMP_API_KEY) {
        throw new Error('FMP_API_KEY not set');
    }
    
    const url = `https://financialmodelingprep.com/api/v3/income-statement/${symbol}?limit=1&apikey=${FMP_API_KEY}`;
    const data = await httpsGet(url);
    
    if (data && data[0] && data[0].netIncome) {
        apiCallCount++;
        return {
            earnings: data[0].netIncome,
            year: new Date(data[0].date).getFullYear(),
            source: 'FMP'
        };
    }
    return null;
}

/**
 * Fetch earnings from Alpha Vantage
 * Free tier: 25 calls/day
 */
async function fetchEarningsFromAlphaVantage(symbol) {
    if (!ALPHA_VANTAGE_API_KEY) {
        throw new Error('ALPHA_VANTAGE_API_KEY not set');
    }
    
    const url = `https://www.alphavantage.co/query?function=INCOME_STATEMENT&symbol=${symbol}&apikey=${ALPHA_VANTAGE_API_KEY}`;
    const data = await httpsGet(url);
    
    if (data && data.annualReports && data.annualReports[0]) {
        apiCallCount++;
        const report = data.annualReports[0];
        return {
            earnings: parseInt(report.netIncome),
            year: new Date(report.fiscalDateEnding).getFullYear(),
            source: 'AlphaVantage'
        };
    }
    return null;
}

/**
 * Fetch earnings from Yahoo Finance (Free, no API key needed)
 * Note: This is less reliable and may require adjustments
 */
async function fetchEarningsFromYahoo(symbol) {
    try {
        const url = `https://query1.finance.yahoo.com/v10/finance/quoteSummary/${symbol}?modules=defaultKeyStatistics,financialData,incomeStatementHistory`;
        const data = await httpsGet(url);
        
        if (data?.quoteSummary?.result?.[0]?.incomeStatementHistory?.incomeStatementHistory?.[0]) {
            const statement = data.quoteSummary.result[0].incomeStatementHistory.incomeStatementHistory[0];
            if (statement.netIncome?.raw) {
                apiCallCount++;
                return {
                    earnings: statement.netIncome.raw,
                    year: new Date(statement.endDate?.raw * 1000).getFullYear(),
                    source: 'Yahoo'
                };
            }
        }
    } catch (error) {
        // Yahoo Finance often blocks automated requests
        console.log(`  Yahoo Finance blocked/failed for ${symbol}`);
    }
    return null;
}

/**
 * Fetch earnings with fallback strategy
 */
async function fetchEarnings(symbol) {
    console.log(`Fetching earnings for ${symbol}...`);
    
    // Try FMP first (most reliable)
    if (FMP_API_KEY) {
        try {
            const result = await fetchEarningsFromFMP(symbol);
            if (result) {
                console.log(`  ✓ Found from FMP: $${(result.earnings / 1e9).toFixed(2)}B`);
                return result.earnings;
            }
        } catch (error) {
            console.log(`  FMP failed: ${error.message}`);
        }
        await sleep(200); // Rate limiting
    }
    
    // Try Alpha Vantage
    if (ALPHA_VANTAGE_API_KEY) {
        try {
            const result = await fetchEarningsFromAlphaVantage(symbol);
            if (result) {
                console.log(`  ✓ Found from Alpha Vantage: $${(result.earnings / 1e9).toFixed(2)}B`);
                return result.earnings;
            }
        } catch (error) {
            console.log(`  Alpha Vantage failed: ${error.message}`);
        }
        await sleep(1000); // Alpha Vantage requires slower rate
    }
    
    // Try Yahoo Finance (free but unreliable)
    try {
        const result = await fetchEarningsFromYahoo(symbol);
        if (result) {
            console.log(`  ✓ Found from Yahoo: $${(result.earnings / 1e9).toFixed(2)}B`);
            return result.earnings;
        }
    } catch (error) {
        console.log(`  Yahoo failed: ${error.message}`);
    }
    
    return null;
}

/**
 * Main function
 */
async function main() {
    console.log('='.repeat(60));
    console.log('EARNINGS DATA FETCHER');
    console.log('='.repeat(60));
    console.log(`Total companies to process: ${capexData.length}`);
    console.log('');
    
    if (!FMP_API_KEY && !ALPHA_VANTAGE_API_KEY) {
        console.error('ERROR: No API keys configured!');
        console.error('Please set at least one of:');
        console.error('  - FMP_API_KEY (Financial Modeling Prep - RECOMMENDED)');
        console.error('  - ALPHA_VANTAGE_API_KEY');
        console.error('');
        console.error('Get free API keys from:');
        console.error('  - https://site.financialmodelingprep.com/developer/docs (250 calls/day)');
        console.error('  - https://www.alphavantage.co/support/#api-key (25 calls/day)');
        process.exit(1);
    }
    
    // Process each company
    for (let i = 0; i < capexData.length; i++) {
        const company = capexData[i];
        
        console.log(`\n[${i + 1}/${capexData.length}] ${company.name} (${company.symbol})`);
        
        // Skip if already has earnings
        if (company.earnings) {
            console.log(`  ⊘ Already has earnings data: $${(company.earnings / 1e9).toFixed(2)}B`);
            companiesProcessed.push(company.symbol);
            continue;
        }
        
        try {
            const earnings = await fetchEarnings(company.symbol);
            
            if (earnings) {
                company.earnings = earnings;
                companiesProcessed.push(company.symbol);
            } else {
                console.log(`  ✗ No earnings data found`);
                companiesFailed.push(company.symbol);
            }
            
            // Rate limiting: pause every 10 requests
            if ((i + 1) % 10 === 0) {
                console.log('\n⏸ Pausing for rate limiting...');
                await sleep(2000);
            }
            
        } catch (error) {
            console.error(`  ✗ Error: ${error.message}`);
            companiesFailed.push(company.symbol);
        }
    }
    
    // Save updated data
    console.log('\n' + '='.repeat(60));
    console.log('SAVING DATA...');
    console.log('='.repeat(60));
    
    fs.writeFileSync('./data/capex_data.json', JSON.stringify(capexData, null, 2));
    fs.writeFileSync('./public/data/capex_data.json', JSON.stringify(capexData, null, 2));
    
    console.log('✓ Saved to ./data/capex_data.json');
    console.log('✓ Saved to ./public/data/capex_data.json');
    
    // Summary
    console.log('\n' + '='.repeat(60));
    console.log('SUMMARY');
    console.log('='.repeat(60));
    console.log(`Total companies: ${capexData.length}`);
    console.log(`✓ Processed: ${companiesProcessed.length}`);
    console.log(`✗ Failed: ${companiesFailed.length}`);
    console.log(`API calls made: ${apiCallCount}`);
    
    if (companiesFailed.length > 0) {
        console.log('\nFailed companies:');
        companiesFailed.forEach(symbol => console.log(`  - ${symbol}`));
    }
    
    console.log('\nDone! ✨');
}

// Run the script
main().catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
});

File: fetch-earnings.py
#!/usr/bin/env python3
"""
Script to fetch earnings data for all companies using Financial Modeling Prep API

RECOMMENDED: Financial Modeling Prep (FMP)
- Free tier: 250 API calls per day
- Most reliable data
- Get free key at: https://site.financialmodelingprep.com/developer/docs

Usage:
    export FMP_API_KEY="your_key_here"
    python3 fetch-earnings.py

Alternative: Use --batch mode to split across multiple days
    python3 fetch-earnings.py --batch 50
"""

import json
import os
import sys
import time
import requests
from typing import Optional, Dict
from pathlib import Path

# Load .env file if it exists
def load_env():
    env_path = Path('.env')
    if env_path.exists():
        with open(env_path) as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith('#') and '=' in line:
                    key, value = line.split('=', 1)
                    os.environ[key] = value

load_env()

# Load configuration
FMP_API_KEY = os.environ.get('FMP_API_KEY', '')
DATA_FILE = './data/financial_data.json'
PUBLIC_DATA_FILE = './public/data/financial_data.json'

# Rate limiting
RATE_LIMIT_DELAY = 0.2  # 200ms between calls (5 calls/second max)
BATCH_SIZE = int(sys.argv[2]) if len(sys.argv) > 2 and sys.argv[1] == '--batch' else None

class EarningsFetcher:
    def __init__(self, api_key: str):
        self.api_key = api_key
        self.api_calls = 0
        self.processed = []
        self.failed = []
        
    def fetch_earnings_fmp(self, symbol: str) -> Optional[Dict]:
        """Fetch earnings from Financial Modeling Prep"""
        if not self.api_key:
            raise ValueError("FMP_API_KEY not set")
        
        url = f"https://financialmodelingprep.com/api/v3/income-statement/{symbol}?limit=1&apikey={self.api_key}"
        
        try:
            response = requests.get(url, timeout=10)
            response.raise_for_status()
            data = response.json()
            
            if data and len(data) > 0 and 'netIncome' in data[0]:
                self.api_calls += 1
                return {
                    'earnings': data[0]['netIncome'],
                    'year': data[0].get('calendarYear', 2024),
                    'source': 'FMP'
                }
        except Exception as e:
            print(f"    FMP API error: {str(e)}")
        
        return None
    
    def process_companies(self, companies: list, start_idx: int = 0, batch_size: Optional[int] = None):
        """Process companies and fetch earnings"""
        
        end_idx = min(start_idx + batch_size, len(companies)) if batch_size else len(companies)
        
        print("=" * 70)
        print("EARNINGS DATA FETCHER - Financial Modeling Prep")
        print("=" * 70)
        print(f"Processing companies {start_idx + 1} to {end_idx} of {len(companies)}")
        print()
        
        if not self.api_key:
            print("ERROR: FMP_API_KEY not set!")
            print()
            print("Get a free API key (250 calls/day) from:")
            print("https://site.financialmodelingprep.com/developer/docs")
            print()
            print("Then set it as an environment variable:")
            print("  export FMP_API_KEY='your_key_here'")
            sys.exit(1)
        
        for i in range(start_idx, end_idx):
            company = companies[i]
            symbol = company['symbol']
            name = company['name']
            
            print(f"[{i + 1}/{len(companies)}] {name} ({symbol})")
            
            # Skip if already has earnings
            if 'earnings' in company and company['earnings']:
                earnings_b = company['earnings'] / 1_000_000_000
                print(f"  ⊘ Already has earnings: ${earnings_b:.2f}B")
                self.processed.append(symbol)
                continue
            
            # Fetch earnings
            try:
                result = self.fetch_earnings_fmp(symbol)
                
                if result:
                    company['earnings'] = result['earnings']
                    earnings_b = result['earnings'] / 1_000_000_000
                    print(f"  ✓ Fetched: ${earnings_b:.2f}B (FY{result['year']})")
                    self.processed.append(symbol)
                else:
                    print(f"  ✗ No data available")
                    self.failed.append(symbol)
                
                # Rate limiting
                time.sleep(RATE_LIMIT_DELAY)
                
            except Exception as e:
                print(f"  ✗ Error: {str(e)}")
                self.failed.append(symbol)
            
            # Progress update every 25 companies
            if (i + 1) % 25 == 0:
                print()
                print(f"Progress: {self.api_calls} API calls used")
                print()
        
        return companies
    
    def save_data(self, companies: list):
        """Save updated data to JSON files"""
        print()
        print("=" * 70)
        print("SAVING DATA...")
        print("=" * 70)
        
        # Save with pretty formatting
        json_data = json.dumps(companies, indent=2, ensure_ascii=False)
        
        with open(DATA_FILE, 'w', encoding='utf-8') as f:
            f.write(json_data)
        print(f"✓ Saved to {DATA_FILE}")
        
        with open(PUBLIC_DATA_FILE, 'w', encoding='utf-8') as f:
            f.write(json_data)
        print(f"✓ Saved to {PUBLIC_DATA_FILE}")
    
    def print_summary(self, total_companies: int):
        """Print summary statistics"""
        print()
        print("=" * 70)
        print("SUMMARY")
        print("=" * 70)
        print(f"Total companies: {total_companies}")
        print(f"✓ Processed: {len(self.processed)}")
        print(f"✗ Failed: {len(self.failed)}")
        print(f"API calls used: {self.api_calls} / 250 daily limit")
        print(f"Remaining: {250 - self.api_calls}")
        
        if self.failed:
            print()
            print("Failed companies (may need manual review):")
            for symbol in self.failed[:10]:  # Show first 10
                print(f"  - {symbol}")
            if len(self.failed) > 10:
                print(f"  ... and {len(self.failed) - 10} more")
        
        print()
        print("Done! ✨")


def main():
    """Main execution"""
    # Load existing data
    try:
        with open(DATA_FILE, 'r', encoding='utf-8') as f:
            companies = json.load(f)
    except FileNotFoundError:
        print(f"Error: {DATA_FILE} not found!")
        sys.exit(1)
    
    # Initialize fetcher
    fetcher = EarningsFetcher(FMP_API_KEY)
    
    # Check if batch mode
    if BATCH_SIZE:
        print(f"Batch mode: Processing {BATCH_SIZE} companies at a time")
        # Find first company without earnings
        start_idx = 0
        for i, company in enumerate(companies):
            if 'earnings' not in company or not company['earnings']:
                start_idx = i
                break
        
        companies = fetcher.process_companies(companies, start_idx, BATCH_SIZE)
    else:
        # Process all companies
        companies = fetcher.process_companies(companies)
    
    # Save data
    fetcher.save_data(companies)
    
    # Print summary
    fetcher.print_summary(len(companies))
    
    # Check if more companies need processing
    remaining = sum(1 for c in companies if 'earnings' not in c or not c['earnings'])
    if remaining > 0:
        print()
        print(f"⚠️  {remaining} companies still need earnings data")
        print(f"Run this script again tomorrow or use --batch mode:")
        print(f"  python3 fetch-earnings.py --batch 50")


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print("\n\nInterrupted by user")
        sys.exit(1)
    except Exception as e:
        print(f"\nFatal error: {str(e)}")
        sys.exit(1)

